AudioDeviceExample/AudioDevices/ExampleCoreAudioDevice.m

```diff
@@ -0,0 +1,487 @@
+    TVIAudioDeviceReadRenderData(context->deviceContext, audioBuffer, audioBufferSizeInBytes);
+    return noErr;
+}
+
+#pragma mark - Private (AVAudioSession and CoreAudio)
+
++ (nullable TVIAudioFormat *)activeRenderingFormat {
+    /*
+     * Use the pre-determined maximum frame size. AudioUnit callbacks are variable, and in most sitations will be close
+     * to the `AVAudioSession.preferredIOBufferDuration` that we've requested.
+     */
+    const size_t sessionFramesPerBuffer = kMaximumFramesPerBuffer;
+    const double sessionSampleRate = [AVAudioSession sharedInstance].sampleRate;
+    const NSInteger sessionOutputChannels = [AVAudioSession sharedInstance].outputNumberOfChannels;
+    size_t rendererChannels = sessionOutputChannels >= TVIAudioChannelsStereo ? TVIAudioChannelsStereo : TVIAudioChannelsMono;
+
+    return [[TVIAudioFormat alloc] initWithChannels:(size_t)rendererChannels
+                                         sampleRate:sessionSampleRate
+                                    framesPerBuffer:sessionFramesPerBuffer];
+}
+
++ (AudioComponentDescription)audioUnitDescription {
+    AudioComponentDescription audioUnitDescription;
+    audioUnitDescription.componentType = kAudioUnitType_Output;
+    audioUnitDescription.componentSubType = kAudioUnitSubType_RemoteIO;
+    audioUnitDescription.componentManufacturer = kAudioUnitManufacturer_Apple;
+    audioUnitDescription.componentFlags = 0;
+    audioUnitDescription.componentFlagsMask = 0;
+    return audioUnitDescription;
+}
+
+- (void)setupAVAudioSession {
+    AVAudioSession *session = [AVAudioSession sharedInstance];
+    NSError *error = nil;
+
+    if (![session setPreferredSampleRate:kPreferredSampleRate error:&error]) {
+        NSLog(@"Error setting sample rate: %@", error);
+    }
+
+    if (![session setPreferredOutputNumberOfChannels:kPreferredNumberOfChannels error:&error]) {
+        NSLog(@"Error setting number of output channels: %@", error);
+    }
+
+    /*
+     * We want to be as close as possible to the 10 millisecond buffer size that the media engine needs. If there is
+     * a mismatch then TwilioVideo will ensure that appropriately sized audio buffers are delivered.
+     */
+    if (![session setPreferredIOBufferDuration:kPreferredIOBufferDuration error:&error]) {
+        NSLog(@"Error setting IOBuffer duration: %@", error);
+    }
+
+    if (![session setCategory:AVAudioSessionCategoryPlayback error:&error]) {
+        NSLog(@"Error setting session category: %@", error);
+    }
+
+    [self registerAVAudioSessionObservers];
+
+    if (![session setActive:YES error:&error]) {
+        NSLog(@"Error activating AVAudioSession: %@", error);
+    }
+
+    if (session.maximumInputNumberOfChannels > 0) {
+        if (![session setPreferredInputNumberOfChannels:TVIAudioChannelsMono error:&error]) {
+            NSLog(@"Error setting number of input channels: %@", error);
+        }
+    }
```

Comment:
Is this required? ..we do not demonstrate recording. so we can say we dont care about `intputNumberOfChannels`.

Reply:
True. We could ditch this.

Reply:
We don't care about it, i'll delete these lines.